src/backend/distributed/commands/call.c

```diff
@@ -217,3 +202,86 @@ CallFuncExprRemotely(CallStmt *callStmt, DistObjectCacheEntry *procedure,
 
 	return true;
 }
+
+
+/*
+ * ShardPlacementWhenColocatedWithReferenceTable decides on a placement for delegating
+ * a procedure call that reads from a reference table.
+ *
+ * If citus.task_assignment_policy is set to round-robin, we assign a different placement
+ * on consecutive runs. Otherwise the function returns the first placement available.
+ */
+static ShardPlacement *
+ShardPlacementWhenColocatedWithReferenceTable(CitusTableCacheEntry *cacheEntry)
+{
+	const ShardInterval *shardInterval = cacheEntry->sortedShardIntervalArray[0];
+	const uint64 referenceTableShardId = shardInterval->shardId;
+	List *placementList = ActiveShardPlacementList(referenceTableShardId);
+
+	if (TaskAssignmentPolicy == TASK_ASSIGNMENT_ROUND_ROBIN)
+	{
+		placementList = RoundRobinReorder(placementList);
+	}
+
+	return (ShardPlacement *) linitial(placementList);
+}
+
+
+/*
+ * ShardPlacementWhenColocatedWithDistTable decides on a placement
+ * for delegating a procedure call that accesses a distributed table.
+ */
+static ShardPlacement *
+ShardPlacementWhenColocatedWithDistTable(DistObjectCacheEntry *procedure,
+										 FuncExpr *funcExpr,
+										 Var *partitionColumn,
+										 CitusTableCacheEntry *cacheEntry)
+{
+	if (procedure->distributionArgIndex < 0 ||
+		procedure->distributionArgIndex >= list_length(funcExpr->args))
+	{
+		ereport(DEBUG1, (errmsg("cannot push down invalid distribution_argument_index")));
+		return NULL;
+	}
+
+	Node *partitionValueNode = (Node *) list_nth(funcExpr->args,
+												 procedure->distributionArgIndex);
+	partitionValueNode = strip_implicit_coercions(partitionValueNode);
+	if (!IsA(partitionValueNode, Const))
+	{
+		ereport(DEBUG1, (errmsg("distribution argument value must be a constant")));
+		return false;
+	}
+
+	Const *partitionValue = (Const *) partitionValueNode;
+	if (partitionValue->consttype != partitionColumn->vartype)
+	{
+		bool missingOk = false;
+
+		partitionValue =
+			TransformPartitionRestrictionValue(partitionColumn, partitionValue,
+											   missingOk);
+	}
+
+	Datum partitionValueDatum = partitionValue->constvalue;
+	ShardInterval *shardInterval = FindShardInterval(partitionValueDatum, cacheEntry);
+	if (shardInterval == NULL)
+	{
+		ereport(DEBUG1, (errmsg("cannot push down call, failed to find shard interval")));
+		return NULL;
+	}
+
+	/* Get the list of active shard placements ordered by the groupid */
+	List *placementList = ActiveShardPlacementList(shardInterval->shardId);
+	placementList = SortList(placementList, CompareShardPlacementsByGroupId);
```

Comment:
What is the reason of sorting the `placementList` if we are going to give an error when there is more than one element?

Reply:
fixed.